src/clandestine.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019. The Tari Project
+// SPDX-License-Identifier: BSD-3-Clause
+
+//! Functionality for securely handling sensitive data in a clandestine way.
+//! Some data, like cryptographic key material, needs special handling.
+//! This kind of data should be zeroized when it goes away, should never be written to log files or displayed, and
+//! should be generally difficult to misuse. This library tackles all three.
+//!
+//! We store the data on the heap and use `zeroize` to ensure it is cleared when it goes out of scope.
+//! We use a `Hidden` wrapper to prevent the data from being displayed or written to debug logs.
+//! We use strict type separation to prevent the data from being used out of context, which is especially important for
+//! key material. And we only allow the data to be accessed by reference and mutable reference, which prevents
+//! unintended moves or copies. You can still clone, but this retains all the guardrails and is safe.
+
+use std::{fmt::Display, marker::PhantomData};
+
+use zeroize::Zeroize;
+
+use crate::Hidden;
+
+/// A marker trait for differentiated clandestine types
+pub trait ClandestineType {}
+
+/// This macro creates a differentiated clandestine type
+#[macro_export]
+macro_rules! clandestine {
+    ($name:ident) => {
+        #[derive(Debug, Eq, PartialEq)]
+        pub struct $name;
+
+        impl ClandestineType for $name {}
+    };
+}
+
+/// The guts of generic clandestine types
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Clandestine<T, K>
+where
+    T: Zeroize + Default,
```

Comment:
Do we need `T` to derive `Zeroize`, though ? What if `T` derives the `Copy` trait, like some `[u8; N]` ?